feat: add SSD tiering docs (#266)
dragonflydb/documentation: docs/managing-dragonfly/tiering.md

---
sidebar_position: 5
---

# SSD Data Tiering in Dragonfly

Dragonfly v1.21.0 introduces a powerful feature: SSD data tiering. This innovative approach optimizes
storage utilization by leveraging SSD/NVMe disks as a secondary storage tier that complements
your existing RAM. By intelligently offloading specific data to faster disk storage,
Dragonfly can significantly reduce physical memory usage, potentially achieving a 2-5x improvement.

## How It Works

> Dragonfly's data tiering focuses on string values exceeding 64 characters in length.

Reviewer: ```suggestion
Dragonfly's data tiering focuses on string values exceeding 64 characters in size.
```
Feels more natural to me, perhaps it's just me

> When enabled, these longer strings are intelligently offloaded to the SSD tier,

Reviewer: You already used the complementing term "intelligently" above, I'd drop it here..

while shorter strings and other data types remain in memory. The main hashtable index is also kept

> in memory, which ensures high performant lookup operations.

Reviewer: ```suggestion
in memory, which ensures high performance lookup operations.
```
I think?

> The offloaded data is still accessible and seamlessly loaded back into RAM when needed for reads.

Reviewer: ```suggestion
The offloaded data is still accessible and seamlessly loaded back into RAM when needed.
```
It's needed for both reads and writes, right? As such, I'd drop this and just say "when needed"

Author: it's not needed for writes. we do not fetch for writes.

Reviewer: idk, it's still confusing. I imagine that for `APPEND`, which to me sounds like a "write" operation, we do read it. no?

Author: yes, for read-modify-write operations we only read it. I will change.

## Enabling Data tiering
The feature can be enable by passing `--tiered_prefix <nvme_path>/<basename>` flag.
Dragonfly will automatically check the free disk space on the partition hosting `<nvme_path>` and

> will deduce the maximum disk space it can use. In order to manually set the maximum

Reviewer: ```suggestion
will deduce the maximum capacity it can use. In order to explicitly set the maximum
```

* **tiered_total_stashes:** - how many offload requests were issued
* **tiered_total_fetches:** - how many times the offloaded items were read from disk
* **tiered_total_deletes:** - how many times the offloaded items were deleted from disk
* **tiered_total_uploads:** - how many times the offloaded items were promoted back to RAM
* **tiered_allocated_bytes:** - how much disk space was used by data tiering.
* **tiered_capacity_bytes:** - the maximum size of tiered on-disk capacity.
* **tiered_pending_read_cnt:**   - currently pending io read requests. A high number indicates
that the server is bottlenecked on disk read i/o.
* **tiered_pending_stash_cnt:**  - currently pending io write requests. A high number indicates that
the server is bottlenecked on disk write i/o.
* **tiered_cold_storage_bytes:** - cooling queue capacity in bytes
* **tiered_ram_hits:**           - how many times an entry lookup resulted in in-memory hit
* **tiered_ram_misses:**         - how many times an entry lookup resulted in a disk read
* **tiered_ram_cool_hits:**      - how many times an entry lookup resulted in cooling buffer hit.

## System requirements

> Dragonfly Data Tiering requires linux kernel version 5.19 or higher with io_uring API enabled.

Reviewer 2: ```suggestion
Dragonfly Data Tiering requires Linux kernel version 5.19 or higher with io_uring API enabled.
```

> **io_uring API is a critical requirement** for SSD Data tiering. Additionally, it demands

Reviewer: ```suggestion
**io_uring API is a critical requirement** for SSD Data tiering. Additionally, it requires
```

> that `<nvme_path>` points to a relatively fast SDD disk. For cloud workloads, using instances

Reviewer: ```suggestion
that `<nvme_path>` points to a relatively fast SSD disk. For cloud workloads, using instances
```

with locally attached SSDs is recommended. See below for specific cloud provider suggestions.

### GCP

> GCP has excellent local ssd hardware that has great performance characteristics.

Reviewer 2: ```suggestion
GCP has excellent local SSD hardware that has great performance characteristics.
```

> Any instance with local ssd storage will suffice.

Reviewer 2: ```suggestion
Any instance with local SSD storage will suffice.
```

### AWS
AWS provides a variety of instance families with local SSD disks attached.
You can refer to the [following table](https://docs.aws.amazon.com/ec2/latest/instancetypes/mo.html#mo_instance-store)
for detailed performance characteristics of local SSDs for each instance type.
We recommend choosing an instance type from the r6gd or m6gd families.

> ## Notes

Reviewer 2: Might be helpful to move the `Notes` section above `How It Works`.

Author: i prefer keeping notes at the end

Data tiering is currently in alpha, which means it's under development and may have
limited functionality or stability. If you encounter any issues while using data tiering,

> please report them by [filing an issue](https://github.com/dragonflydb/dragonfly/issues/)

Reviewer 2: ```suggestion
please report them by [filing an issue](https://github.com/dragonflydb/dragonfly/issues/).
```

* **tiered_pending_read_cnt:**   - currently pending io read requests. A high number indicates
that the server is bottlenecked on disk read i/o.
* **tiered_pending_stash_cnt:**  - currently pending io write requests. A high number indicates that
the server is bottlenecked on disk write i/o.
* **tiered_cold_storage_bytes:** - cooling queue capacity in bytes
* **tiered_ram_hits:**           - how many times an entry lookup resulted in in-memory hit
* **tiered_ram_misses:**         - how many times an entry lookup resulted in a disk read
* **tiered_ram_cool_hits:**      - how many times an entry lookup resulted in cooling buffer hit.

## Performance
Performance benchmarks against Elasticache instances and Memcached,
conducted on AWS instances, demonstrate Dragonfly's superior performance.
We've conducted loadtests using `r6gd.xlarge` instance on AWS and compared Dragonfly against
datastore with similar features, namely Elasticache Data tiering `cache.r6gd.xlarge` instance and
self-hosted Memcached/ExtStore server also running on `r6gd.xlarge`. The test consisted of
writing a 90GB dataset into all stores and then reading them randomly with uniform distribution.

> During the write phase with 200K rps, Memcached had to drop 20% of the workload in order

Reviewer: I'd capitalize RPS throughout, but your call

## System requirements
Dragonfly Data Tiering requires Linux kernel version 5.19 or higher with io_uring API enabled.
**io_uring API is a critical requirement** for SSD Data tiering. Additionally, it requires
that `<nvme_path>` points to a relatively fast SSD disk. For cloud workloads, using instances
with locally attached SSDs is recommended. See below for specific cloud provider suggestions.

### GCP
GCP has excellent local SSD hardware that has great performance characteristics.
Any instance with local SSD storage will suffice.

### AWS
AWS provides a variety of instance families with local SSD disks attached.
You can refer to the [following table](https://docs.aws.amazon.com/ec2/latest/instancetypes/mo.html#mo_instance-store)
for detailed performance characteristics of local SSDs for each instance type.
We recommend choosing an instance type from the r6gd or m6gd families.

## Notes
Data tiering is currently in alpha, which means it's under development and may have
limited functionality or stability. If you encounter any issues while using data tiering,
please report them by [filing an issue](https://github.com/dragonflydb/dragonfly/issues/).

**Limitations:**

> * Data tiering is not currently supported by BITOP operations.

Reviewer: Probably also HLL right?
